### PR TITLE
[X86][AMX] Support AMX-TF32

### PR DESCRIPTION
**Pull request by fzou1**
Ref.: https://cdrdv2.intel.com/v1/dl/getContent/671368

### CODE REVIEW

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 1]
```diff
@@ -37686,6 +37686,28 @@ X86TargetLowering::EmitInstrWithCustomInserter(MachineInstr &MI,
...
+      break;
+    }
+    MachineInstrBuilder MIB = BuildMI(*BB, MI, DL, TII->get(Opc));
+    MIB.addReg(TMMImmToTMMReg(MI.getOperand(0).getImm()), RegState::Define);
+    MIB.addReg(TMMImmToTMMReg(MI.getOperand(0).getImm()), RegState::Undef);
+    MIB.addReg(TMMImmToTMMReg(MI.getOperand(1).getImm()), RegState::Undef);
+    MIB.addReg(TMMImmToTMMReg(MI.getOperand(2).getImm()), RegState::Undef);
+    MI.eraseFromParent();
+    return BB;
+  }
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:37710]: Merge to line 37473

File: llvm/test/CodeGen/X86/amx-tf32-internal.ll [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  %c1 = call x86_amx @llvm.x86.tmmultf32ps.internal(i16 8, i16 8, i16 8, x86_amx %c, x86_amx %a, x86_amx %b)
+  %c2 = call x86_amx @llvm.x86.ttmmultf32ps.internal(i16 8, i16 8, i16 8, x86_amx %c1, x86_amx %a, x86_amx %b)
+
+  call void @llvm.x86.tilestored64.internal(i16 8, i16 8, i8* %pointer, i64 %stride, x86_amx %c2)
+  ret void
+}
+
+declare x86_amx @llvm.x86.tilezero.internal(i16, i16)
+declare x86_amx @llvm.x86.tileloadd64.internal(i16, i16, i8*, i64)
+declare x86_amx @llvm.x86.tileloaddt164.internal(i16, i16, i8*, i64)
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/amx-tf32-internal.ll:42]: Not used.

File: clang/lib/Headers/amxtf32intrin.h [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+#ifndef __AMX_TF32INTRIN_H
+#define __AMX_TF32INTRIN_H
+#ifdef __x86_64__
+
+#define __DEFAULT_FN_ATTRS_TF32                                                \
+  __attribute__((__always_inline__, __nodebug__, __target__("amx-tf32")))
+
+#define __DEFAULT_FN_ATTRS_TF32_TRANSPOSE                                      \
+  __attribute__((__always_inline__, __nodebug__,                               \
+                 __target__("amx-tf32,amx-transpose")))
```
Comment (phoebewang) [inline clang/lib/Headers/amxtf32intrin.h:23]: We should split transpose intrinsics to a seperate file with macro
```
#if !defined(__SCE__) || __has_feature(modules) ||
   (defined(__AMX_TF32__) && defined(__AMX_TRANSPOSE__))
#include <amxtf32transposeintrin.h>
#endif
```

File: clang/lib/Headers/CMakeLists.txt [hunk 4]
```diff
@@ -151,6 +151,7 @@ set(x86_files
   amxfp16intrin.h
   amxfp8intrin.h
   amxintrin.h
+  amxtf32intrin.h
```
Comment (phoebewang) [inline clang/lib/Headers/CMakeLists.txt:154]: Missing the new file.

File: clang/lib/Headers/immintrin.h [hunk 5]
```diff
@@ -660,6 +660,10 @@ _storebe_i64(void * __P, long long __D) {
 #include <amxavx512intrin.h>
 #endif
 
+#if !defined(__SCE__) || __has_feature(modules) || defined(__AMX_TF32__)
+#include <amxtf32intrin.h>
+#endif
+
```
Comment (phoebewang) [inline clang/lib/Headers/immintrin.h:666]: ditto.

File: llvm/include/llvm/IR/IntrinsicsX86.td [hunk 6]
```diff
@@ -6101,6 +6101,25 @@ let TargetPrefix = "x86" in {
               Intrinsic<[llvm_v16i32_ty],
                         [llvm_i16_ty, llvm_i16_ty, llvm_x86amx_ty, llvm_i32_ty],
                         []>;
+
+  def int_x86_tmmultf32ps : ClangBuiltin<"__builtin_ia32_tmmultf32ps">,
+              Intrinsic<[], [llvm_i8_ty, llvm_i8_ty, llvm_i8_ty],
+              [ImmArg<ArgIndex<0>>, ImmArg<ArgIndex<1>>, ImmArg<ArgIndex<2>>]>;
+  def int_x86_ttmmultf32ps : ClangBuiltin<"__builtin_ia32_ttmmultf32ps">,
+              Intrinsic<[], [llvm_i8_ty, llvm_i8_ty, llvm_i8_ty],
+              [ImmArg<ArgIndex<0>>, ImmArg<ArgIndex<1>>, ImmArg<ArgIndex<2>>]>;
```
Comment (phoebewang) [inline llvm/include/llvm/IR/IntrinsicsX86.td:6110]: Align within `<`

File: clang/lib/Headers/CMakeLists.txt (hunk 4 above)
Comment (fzou1) [reply clang/lib/Headers/CMakeLists.txt:154]: Sorry. Forgot to add it. Done. Thanks.

File: clang/lib/Headers/immintrin.h (hunk 5 above)
Comment (fzou1) [reply clang/lib/Headers/immintrin.h:666]: Added.

File: llvm/include/llvm/IR/IntrinsicsX86.td (hunk 6 above)
Comment (fzou1) [reply llvm/include/llvm/IR/IntrinsicsX86.td]: Done

**Review by phoebewang — APPROVED** [review]
LGTM.

**Pull request merged**